### PR TITLE
Do not use zero(...) in generic_lu! algorithm to support Unitful

### PR DESCRIPTION
**Pull request by haampie**
This is a minimal effort attempt to work around the bigger problem that factorizations only define a single number type.

Suppose `eltype(A) == typeof(1.0u"kg")`, then by design of the algorithm the LU factorization should end up with `A = L * U` where `eltype(L) = Float64` and `eltype(U) == eltype(A) == typeof(1.0u"kg")`.

In julia the algorithm assumes it can work in-place, so we have to promote/union number types of L and U:

```
julia> promote_type(Float64, typeof(1.0u"kg"))
Quantity{Float64, D, U} where U where D
```

and in this case `zero(promote_type(Float64, typeof(1.0u"kg")))` will error. This PR tries to avoid using zero then.

A better solution IMHO is to allocate L and U separately (or at the very least parametrize the eltype of L and U separately in the factorization) to avoid having to work with non-concrete number types. It's nice to get a unitful LU-factorization, but pretty much any linalg operation `\`, `*`, `+`, etc will run `oneunit(eltype(factorization))` and throw:

```
julia> F = lu(rand(typeof(1.0u"kg"), 3, 3))
LU{Quantity{Float64, D, U} where U where D, Matrix{Quantity{Float64, D, U} where U where D}}
L factor:
3×3 Matrix{Quantity{Float64, D, U} where U where D}:
      1.0     0.0 kg  0.0 kg
 0.268046        1.0  0.0 kg
 0.512366  -0.566135     1.0
U factor:
3×3 Matrix{Quantity{Float64, D, U} where U where D}:
 0.654136 kg  0.762665 kg  0.637124 kg
         0.0  0.640381 kg  0.682195 kg
         0.0          0.0  0.102111 kg

julia> F \ rand(3)
ERROR: MethodError: no method matching (Quantity{Float64, D, U} where U where D)(::Float64)
Closest candidates are:

```

### CODE REVIEW

**Review by stevengj — APPROVED** [review]

File: stdlib/LinearAlgebra/src/lu.jl [hunk 1]
```diff
@@ -139,9 +139,9 @@ function generic_lufact!(A::StridedMatrix{T}, ::Val{Pivot} = Val(true);
         for k = 1:minmn
             # find index max
             kp = k
-            if Pivot
-                amax = abs(zero(T))
-                for i = k:m
+            if Pivot && k < m
```
Comment (Keno) [inline stdlib/LinearAlgebra/src/lu.jl:142]: Why not `k <= m`? Otherwise this is not a no-op.

**Comment by Keno** [comment]
Could use a test. We have `Furlong` in the test code for this purpose.

File: stdlib/LinearAlgebra/src/lu.jl (hunk 1 above)
Comment (haampie) [reply stdlib/LinearAlgebra/src/lu.jl:142]: Both are correct :) if `k == m` there's only one element left in `A[k:m, k]`, so you don't have to search for the maximum, it's `k`, which is already set in line 141.

**Comment by Keno** [comment]
This hit during CI downtime, but otherwise LGTM. Should have a test though, the addition of which will also re-trigger CI.

**Comment by haampie** [comment]
I think we can't use Furlong here, since * is not a closed operation. It would mean the eltype would depend on the indices, meaning that the matrix would have some abstract type, in which case simple stuff like one(T) won't work.

**Comment by Keno** [comment]
I'm confused what this change does then. Isn't that the same issue with Unitful also?

**Comment by haampie** [comment]
Sorry for the confusion, it's exactly the same as Unitful yes. There was an error displaying the LU-factorization of Furlongs in the REPL where `one(T)` and `zero(T)` were used, which made me think the computation failed, but in fact it works fine. Somehow this printing error doesn't occur for Unitful.

**Comment by dkarrasch** [comment]
The only failure is a timeout, but the LU tests passed. So this is ready to go, right?

**Pull request merged**